apisix/plugins/ocsp-stapling.lua

```diff
@@ -0,0 +1,253 @@
+    if not ok then
+        return false, "failed to set PEM priv key: " .. err
+    end
+
+    return true
+end
+
+
+local function fetch_ocsp_resp(der_cert_chain)
+    core.log.info("fetch ocsp response from remote")
+    local ocsp_url, err = ngx_ocsp.get_ocsp_responder_from_der_chain(der_cert_chain)
+
+    if not ocsp_url then
+        -- if cert not support ocsp, the report error is nil
+        if not err then
+            err = "cert not contains authority_information_access extension"
+        end
+        return nil, "failed to get ocsp url: " .. err
+    end
+
+    local ocsp_req, err = ngx_ocsp.create_ocsp_request(der_cert_chain)
+    if not ocsp_req then
+        return nil, "failed to create ocsp request: " .. err
+    end
+
+    local httpc = http.new()
+    local res, err = httpc:request_uri(ocsp_url, {
+        method = "POST",
+        headers = {
+            ["Content-Type"] = "application/ocsp-request",
+        },
+        body = ocsp_req
+    })
+
+    if not res then
+        return nil, "ocsp responder query failed: " .. err
+    end
+
+    local http_status = res.status
+    if http_status ~= 200 then
+        return nil, "ocsp responder returns bad http status code: "
+               .. http_status
+    end
+
+    if res.body and #res.body > 0 then
+        return res.body, nil
+    end
+
+    return nil, "ocsp responder returns empty body"
+end
+
+
+local function set_ocsp_resp(full_chain_pem_cert, skip_verify, cache_ttl)
+    local der_cert_chain, err = ngx_ssl.cert_pem_to_der(full_chain_pem_cert)
+    if not der_cert_chain then
+        return false, "failed to convert certificate chain from PEM to DER: ", err
+    end
+
+    local ocsp_resp = ocsp_resp_cache:get(der_cert_chain)
+    if ocsp_resp == nil then
+        core.log.info("not ocsp resp cache found, fetch from ocsp responder")
+        ocsp_resp, err = fetch_ocsp_resp(der_cert_chain)
+        if ocsp_resp == nil then
+            return false, err
+        end
+        core.log.info("fetch ocsp resp ok, cache it")
+        ocsp_resp_cache:set(der_cert_chain, ocsp_resp, cache_ttl)
+    end
+
+    if not skip_verify then
+        local ok, err = ngx_ocsp.validate_ocsp_response(ocsp_resp, der_cert_chain)
+        if not ok then
+            return false, "failed to validate ocsp response: " .. err
+        end
+    end
+
+    -- set the OCSP stapling
+    local ok, err = ngx_ocsp.set_ocsp_status_resp(ocsp_resp)
+    if not ok then
+        return false, "failed to set ocsp status response: " .. err
+    end
+
+    return true
+end
+
+
+local original_set_cert_and_key
+local function set_cert_and_key(sni, value)
+    if value.gm then
+        -- should not run with gm plugin
+        core.log.info("gm plugin enabled, no need to run ocsp-stapling plugin")
+        return original_set_cert_and_key(sni, value)
+    end
+
+    if not value.ocsp_stapling then
+        core.log.info("no 'ocsp_stapling' field found, no need to run ocsp-stapling plugin")
```

Comment:
ditto

Reply:
It will execute the original function `set_cert_and_key()`, same as `return original_set_cert_and_key(sni, value)`.